workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt

```diff
@@ -0,0 +1,180 @@
+     * [ViewTreeLifecycleOwner], if that fails it looks up the context chain for a [LifecycleOwner],
+     * and if that fails it just returns null.
+     */
+    private fun findViewTreeLifecycle(view: View): Lifecycle? {
+      // Start at our view's parent – if we look on our own view, we'll just get this back.
+      return (view.parent as? View)?.let { ViewTreeLifecycleOwner.get(it) }?.lifecycle
+        ?: view.context.lifecycleOrNull()
+    }
+  }
+}
+
+/**
+ * @param enforceMainThread Allows disabling the main thread check for testing.
+ */
+@OptIn(WorkflowUiExperimentalApi::class)
+@VisibleForTesting(otherwise = PRIVATE)
+internal class RealWorkflowLifecycleOwner(
+  private val view: View,
+  private val findParentLifecycle: () -> Lifecycle?,
+  enforceMainThread: Boolean = true,
+) : WorkflowLifecycleOwner,
+  LifecycleOwner,
+  OnAttachStateChangeListener,
+  LifecycleEventObserver {
+
+  private val localLifecycle =
+    if (enforceMainThread) LifecycleRegistry(this) else createUnsafe(this)
+  private var parentLifecycle: Lifecycle? = null
+  private var destroyOnDetach = false
+
+  override fun onViewAttachedToWindow(v: View) {
+    val oldLifecycle = parentLifecycle
+    parentLifecycle = findParentLifecycle()
+
+    if (parentLifecycle !== oldLifecycle) {
+      oldLifecycle?.removeObserver(this)
+      parentLifecycle?.addObserver(this)
+    }
+    updateLifecycle(isAttached = true)
+  }
+
+  override fun onViewDetachedFromWindow(v: View) {
+    updateLifecycle(isAttached = false)
+  }
+
+  /** Called when the [parentLifecycle] changes state. */
+  override fun onStateChanged(
+    source: LifecycleOwner,
+    event: Event
+  ) {
+    updateLifecycle()
+  }
+
+  override fun destroyOnDetach() {
+    if (!destroyOnDetach) {
+      destroyOnDetach = true
+      updateLifecycle()
+    }
+  }
+
+  override fun getLifecycle(): Lifecycle = localLifecycle
+
+  /**
+   * @param isAttached Whether the view is [attached][View.isAttachedToWindow]. Must be passed
+   * explicitly when called from the attach/detach callbacks, since the view property's value won't
+   * reflect the new state until after they return.
+   */
+  @VisibleForTesting(otherwise = PRIVATE)
+  internal fun updateLifecycle(isAttached: Boolean = view.isAttachedToWindow) {
+    val parentState = parentLifecycle?.currentState
+    val localState = localLifecycle.currentState
+
+    if (localState == DESTROYED) {
+      // Local destruction is a terminal state.
+      return
+    }
+
+    localLifecycle.currentState = when {
+      // We've been queued for destruction.
+      // Stay attached to the parent's lifecycle until we re-attach, since the parent could be
+      // destroyed while we're detached.
+      destroyOnDetach && !isAttached -> DESTROYED
+      // We may or may not be attached, but we have a parent lifecycle so we just blindly follow it.
+      parentState != null -> parentState
```

Comment:
I think this is too permissive. We should crash if there's no parent lifecycle, since the root of the lifecycle tree should always be a `Fragment` or `ComponentActivity`. If it's not, we might leak all over the place.

Reply:
When exactly should we crash? You don't want to require getting attached, so parent lifecycle really can be null forever, right?

Reply:
The following invariants should hold:
- If we have ever been attached, and we have not been destroyed, then there must be a parent lifecycle.
- If there is no parent lifecycle, then either we must be in the INITIALIZED state (and never been attached), OR we must be in the DESTROYED state.

Reply:
That'll be a good comment.

Reply:
It'll be an even better set of unit tests.

Reply:
Done.